README.md
# Quake Stats
[![PyPI](https://img.shields.io/pypi/v/quakestats)](https://pypi.org/project/quakestats/)
[![PyPI - Wheel](https://img.shields.io/pypi/wheel/quakestats)](https://pypi.org/project/quakestats/#files)

Quake 3 logs / Quake Live events processing project.

Allows to retrieve, process, analyze, store and present Quake matches.

The project doesn't aim to give global stats like [qlstats](http://qlstats.net) it's rather meant to store statistics from some server group (server domain). The origins of Quake Stats come from a group of players who occasionally play together and want to keep track of their matches... and to have fun from some custom made medals (badges) :)

## Overview
### Supported features:
* processing Quake 3 logs (log parsing, transforming to QL)
* processing Quake Live event streams (zmq listen on QL server stat endpoint)
* translating (to some extent) Quake 3 logs into Quake Live events
* analysing matches
* storing matches in Database backend (Mongo DB)
* presenting match results through a web application

### Supported mods and game modes
Unfortunately only OSP FFA from Quake 3 is well tested as it was the main use case
#### mods
- [x] - OSP (http://www.orangesmoothie.org/tourneyQ3A/index.html)
- [x] - Quake Live - most of event processing is implemented
- [x] - Edawn - requires 1.6.3+ (enchanced logging)
- [ ] - vanilla Q3 not supported (need workaround for missing ServerTime)
- [ ] - CPMA not supported (need workaround for missing ServerTime)

#### modes
- [x] - DUEL
- [x] - FFA
- [ ] - CA - partially implemented
- [ ] - TDM
- [ ] - CTF

### Custom medals
Are described here [resources.js](quakestats/static/js/resources.js)

### Examples
The stats are presented with fancy charts, custom medals, etc. See the examples below.
#### Total badges/medals board
![home1](examples/home1.png)
#### Total kills & deaths
![home2](examples/home2.png)
#### Single match Kill Death Ratio, Worst Enemy, Score over Time chart
![match1](examples/match1.png)

### Requirements
- Python 3.6+
- Instance of Mongo DB (pointed by ```settings.py```)
- Modern web browser (requires css grid-layout)

## How to setup
In order to setup the application you need to have python 3.6+ (virtualenv recommended) and an instance of mongo DB.

### Installation
#### Install from pip package
```bash
pip install quakestats
```

#### Install from source code (optional)
Is also needed install ```quakestats``` package (in virtualenv if you are using it). To do that you could install it directly
```bash
pip install -r requirements.txt
python setup.py install
```

### Configuration file
The application is configured by setting ```QUAKESTATS_SETTINGS``` environment variable to path to configuration file.
See example [settings.py](examples/settings.py)

### Verify if everything is properly set up
Quake Stats provide a simple CLI with a command to verify an environment
```bash
quakestats status
```

Example output:
```
(venv) [user@pc quakestats]$ quakestats status
app -> version: 0.9.61
settings -> env var: /opt/quakestats/settings.py
settings -> RAW_DATA_DIR: /opt/quakestats/data
db -> ping: {'ok': 1.0}
webapp -> loadable: Quakestats webapp is loadable
```

### Run Quake Stats web app
You can setup Quake Stats web app with any websever you want (as long as it supports python, e.g. mod wsgi, uwsgi).
This documentation covers only running in *twisted* webserver

#### Run in twistd (example)
You can launch Quake Stats web application using ```twistd``` webserver. Just make sure to install twisted framework first.
Also make sure to use some recent version of twisted (tested with 18.7.0 installed by pip).
```bash
pip install twisted
FLASK_APP="quakestats.web"; QUAKESTATS_SETTINGS=`pwd`/settings.py; twistd web --wsgi quakestats.web.app
```

## User/Admin guide
### Setup admin user
Admin user is used by web application to access some additional administrative operations. For now it's only setting map sizes. Just to have a list of recently used maps and their sizes. Nothing more at the moment.
```bash
# you need to run the command in proper python environment
# use "quakestats status" to check your environment
quakestats set-admin-pwd <yourpassword>
```

### Collecting Quake Live stats
Quake Live exposes stats server through tcp socket (zmq) authenticated with password.
CLI can gather stats from multiple QL servers and process them automatically.
Use following config file

```ini
[server-1]
ip = 5.6.7.8
port = 27967
password = password1

[server-1]
ip = 1.2.3.4
port = 27967
password = password2

```

Use following CLI to start collecting events (assuming your config file is named ```collector.cfg```)

```bash
quakestats collect-ql collector.cfg
```

### Uploading Quake 3 log file
In order to process some data you need to send your match log file to web api endpoint ```/api/v2/upload```. By default mod ```osp``` is assumed.
Mod specific endpoint is served under ```/api/v2/upload/<mod>```, e.g. ```/api/v2/upload/edawn```
You need an ```ADMIN_TOKEN``` set in configuration.
```bash
curl -X POST --form file=@/path/to/your/games.log --form token=adminsecrettoken host:port/api/v2/upload
```

To send vanila q3 log use
```
curl -X POST --form file=@/path/to/your/games.log --form token=adminsecrettoken host:port/api/v2/upload/baseq3
```

To send edawn log use
```
curl -X POST --form file=@/path/to/your/games.log --form token=adminsecrettoken host:port/api/v2/upload/edawn

```
All log files with extracted matches are stored in directory determined by ```RAW_DATA_DIR``` config entry

### Using automated script to send logs
TODO, deprecated

### Rebuilding database
You can rebuild your database using files stored in ```RAW_DATA_DIR``` with simple web api call or CLI.
```bash
curl -X POST host:port/api/v2/admin/rebuild --form token=adminsecrettoken
```
```bash
# you need to run the command in proper python environment
# use "quakestats status" to check your environment
quakestats rebuild-db
```
If you implement some new Medals or any other backend related feature this API call will clear previous data stored in DB and process all matches from data directory once again.

### Merging player results
Unfortunately the only way to distinguish players in Quake 3 servers is to use player nickname. When player changes his nickname between matches he will be treated as new unique player. In such cases admin can merge results of two specific players. Use with caution as it will rewrite history of all matches stored in database.
```bash
curl -X POST host:port/api/v2/admin/players/merge --form token=admintoken --form source_player_id=297f6272f79d4918c4efe098 --form target_player_id=df55e5cd4582d6f14cd20746
```
It will merge all results from player with id ```297f6272f79d4918c4efe098``` into player with id ```df55e5cd4582d6f14cd20746```. To find out how player ID is build see the development section.

### Importing preprocessed match log
Preprocessed match logs stored in ```RAW_DATA_DIR``` can be imported using admin match import API.
This can be particularly useful when e.g. debugging some bugs on dev infra.
```bash
curl -X POST --form file=@bugmatch.log --form token=admintoken host:port/api/v2/admin/match/import
```

## Development
### Tech stack
Python, Flask, MongoDB, d3.js, riot.js, zmq

##### How does it work with Quake 3 Players
Quake 3 players don't have unique ID's so it's hard to distinguish players between matches. In order to overcome this problem each player has ```player_id``` assigned during match analysis. The ID is constructed as hash of ```SERVER_DOMAIN``` and player nickname as a result it's consistent between matches as long as player keeps his nickname and there is no nickname clash. Perhaps there is some better way? Server side auth?

#### Web
Web application related components
- api - web API used by frontend and to retrieve Quake 3 logs
- views - typical flask views

### Extending
#### How to add new medal
- see [SpecialScores class](quakestats/dataprovider/analyzer/specials.py) - for special scores
- see [Badger class](quakestats/dataprovider/analyzer/badges.py) - for badges calculation
- see [JS resources](quakestats/static/js/resources.js) - to add new medal image

#### Running tests
```bash
make test
```
### Assets
Medals, icons, etc.
Some of the assets are missing it would be nice to find some free ones or draw them ;)

### How to release new version
```bash
bumpversion <major|minor|patch> --commit --tag
```

## Troubleshooting
### Can't use mongodb 5.0+ without AVX on Ubuntu 20+
This problem occurs on older hardware when mongodb 5.0+ is being used.
You can't use official mongodb packages with version higher than 4.x. 4.x series is the last one not using AVX. BUT it requires libssl1.1 and it's no longer shipped with Ubuntu.
Fortunately it's possible to install libssl1.1 from http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/. Then just install mongodb 4.x.


CHANGELOG.MD
Changelog
=========

NEXT RELEASE
-------------------
- Feature: Support baseq3 vanila logs
- Code: Fixed dependencies for Python3.10
- Code: Bumped flask version

v1.5.1 (2021-10-17)
-------------------
- Feature: Added special score for taking Quad Damage
- Feature: Blue star appears on score chart when player picks Quad Damage

v1.5.0 (2020-11-20)
-------------------
- Feature: Proper support for QL and Q3 logs in the same instance
- Feature: Added Quake3 edawn mod support (requires enchanced logging, not yet released)
- Feature: Added QL stats collector (single process can collect from multiple servers)
- Bugfix: aborted QL matches will be ignored
- Code: Rearranged project according to PYPA
- Code: Bumped Python minimal version to 3.6 (typehints, asyncio)
- Code: Rewritten raw data dir as log warehouse
- Code: Rewritten a lot of code as SDK (CLI/WEB -> SDK)
- Code: Added system context, module/component API classes
- Code: Reorganization of parser code


v1.4.0 (2020-04-21)
-------------------
- Fixed QL store to file. [BR]
- QL collector will process matches right after match received. [BR]
- Better names in kills-deaths details. [BR]
- Better player colors, initial work for fancy player names. [BR]
- Added receive time to game metadata. [BR]
- QL collector will try to reconnect if there are no events for 5m. [BR]
- Added CLI for listening for QL server stats. [BR]
- Added better weapon stats panel. [BR]
- Renamed .tag to .riot. [BR]
- More q3 MOD_ weapon support, added game finish date. [BR]
- Endpoint to fetch total stats. [BR]
- Bugfix: on merge players rewrite stats as well. [BR]
- Code reorganization in tests [BR]
- Added flake8 to CI. [BR]
- Added checksum calculation. [BR]
- Rewritten q3toql. [BR]
- Fixed webapp resources path for packaging. [BR]
- Moved around web application code. [BR]
- Added flake8, commit hook, fixed linter errors. [BR]
- Applied black on codebase. [BR]
- Appplied isort on whole codebase. [BR]


v1.3.1 (2020-02-05)
-------------------
- Fixed silent UI bug when weapon accuracy stat is missing. [BR]


v1.3.0 (2020-02-05)
-------------------

Fix
~~~
- 'VS' is now png. [BR]

Other
~~~~~
- Slightly adjusted layout, added accuracy info. [BR]
- Fixed tests. [BR]
- Removed broken tr. [BR]
- Use wretch.js for presence list. [BR]
- Added accuracy info for duels. [BR]
- Dev: Added 'requests' module requirement. [BR]
- Dev: Fixed tests. [BR]
- Dev: Renamed analysis report description to summary. [BR]
- Ignore 'late' events when calculating scores. [BR]
- Add new q3toql event attribute to distinguish 'late' events. [BR]
- Added description to duel matches. [BR]
- Bugfix: inactive player is no longer added to the score log. [BR]
- Add proper duel player names. [BR]
- Added view for showing 1v1 matches. [BR]
- Log watcher will consume last line during 2nd read. [BR]
- Dev: Fixed tests. [BR]
- Modified main page layout, added all matches list. [BR]
- Bugfix: q3 log watcher can consume half-written log entry. [BR]


v1.2.0 (2019-11-24)
-------------------
- Dev: Fixed tests. [BR]
- Docs: Added note about q3 log watch script. [BR]
- Feature: Added CLI script to watch Q3 log file and send logs
  automatically. [BR]
- Dev: added some logging infra code. [BR]
- Change: Added backend API to accept q3 match logs as json. [BR]
- Change: 'WIN' badges are now assigned according to number of players.
  [BR]
- Bugfix: when fraglimit reached last kill was incorrectly removed from
  scores. [BR]
- Fixed style for changelog generation. [BR]


v1.1.0 (2019-11-19)
-------------------
- Improved changelog generation. [BR]
- Total stats GUI is able to fetch last N matches. [BR]
- Added changelog, makefile changelog script. [BR]
- Added api for fetching badges/total stats from last N matches. [BR]


v1.0.1 (2019-11-18)
-------------------
- Improved minor images, small css fixes. [BR]


v1.0.0 (2019-11-17)
-------------------
- Fixed tests. [BR]
- Added basic presence list. Fix #48. [BR]
- Create pythonpublish.yml. [BR]
- Added pypi badges to readme. [BR]


v0.13.0 (2019-11-17)
--------------------
- Improved installation guide (pip included) [BR]
- Added proper bumpversion config. [BR]
- Don't add 0 score for specting clients. [BR]
- Fixed KDR calculation for non playing clients. [BR]
- Fixed bug when 0 badges can be assigned, improved KDR calculation.
  [BR]


v0.12.0 (2019-10-01)
--------------------
- Fixed mismatched version. [BR]
- Improved top badgers panel. [BR]
- Added suicide bomber badge, Fix #46. [BR]
- Slightly improved badges descriptions. [BR]
- Added consecutive rail kill badge, Fix #34. [BR]
- Improved number of badges visible on homepage. [BR]
- Added badge images for suicide bomber and consecutive rail. [BR]


v0.11.0 (2019-08-25)
--------------------
- Added Marauder badge, Fix #43. [BR]


v0.10.0 (2019-06-09)
--------------------
- Further cleanup in readme file. [BR]
- Improved readme file. [BR]
- Improved test data discovery, use native py3 mock. [BR]
- Added initial CLI, moved around some code. [BR]
- Updated readme, added dev requirements. [BR]


v0.9.61 (2019-05-12)
--------------------
- Added printable option for player view. [BR]
- Added basic frontend support for health/armor/damage stats. [BR]
- Added more logging and exception handling. [BR]
- Added full backend support for match player stats. [BR]
- Added initial support for Weapon_Stats log entry. [BR]
- Moved some test code. [BR]
- Added lumberjack medal. [BR]
- Added lumberjack medal (consecutive gauntlet kills) [BR]
- Added new badge - Ghost Kill. [BR]
- Added match end indicator on score chart Fix #38. [BR]
- Added color generation by player id - Fix #37. [BR]
- Selected player is easier to see now. [BR]
- Its possible to select all player occurance on match page. [BR]
- Added kamikaze badge Fix #35. [BR]
- Code cleanup, Use events instead of dicts. [BR]
- Unified the way test data is generated. [BR]
- Refactored player switch team event. [BR]

quakestats/web/api.py
import logging
from collections import (
    defaultdict,
)

import flask

from quakestats import (
    manage,
)
from quakestats.web.app import (
    app,
    data_store,
)

logger = logging.getLogger("quakestats.webapp")


def auth(token):
    return token == app.config["ADMIN_TOKEN"]


@app.route("/api/v2/matches")
def api2_matches():
    return flask.jsonify(data_store().get_matches(25))


@app.route("/api/v2/matches/all")
def api2_matches_all():
    return flask.jsonify(data_store().get_matches())


@app.route("/api/v2/match/<match_guid>/metadata")
def api2_match_metadata(match_guid):
    return flask.jsonify(data_store().get_match_metadata(match_guid))


@app.route("/api/v2/match/<match_guid>/players")
def api2_match_players(match_guid):
    return flask.jsonify(data_store().get_match_players(match_guid))


@app.route("/api/v2/match/<match_guid>/teams")
def api2_match_team_lifecycle(match_guid):
    return flask.jsonify(data_store().get_team_lifecycle(match_guid))


@app.route("/api/v2/player/<player_id>")
def api2_player(player_id):
    return flask.jsonify(data_store().get_player(player_id))


@app.route("/api/v2/match/<match_guid>/score")
def api2_match_scores(match_guid):
    return flask.jsonify(data_store().get_match_scores(match_guid))


@app.route("/api/v2/match/<match_guid>/special")
def api2_match_special(match_guid):
    return flask.jsonify(data_store().get_match_special_scores(match_guid))


@app.route("/api/v2/match/<match_guid>/kill")
def api2_match_kill(match_guid):
    return flask.jsonify(data_store().get_match_kills(match_guid))


@app.route("/api/v2/match/<match_guid>/badge")
def api2_match_badge(match_guid):
    return flask.jsonify(data_store().get_match_badges(match_guid))


@app.route("/api/v2/match/<match_guid>/player_stats")
def api2_match_player_stats(match_guid):
    return flask.jsonify(data_store().get_match_player_stats(match_guid))


@app.route("/api/v2/board/badges")
def api2_board_badges():
    latest = flask.request.args.get("latest", default=None)
    latest = int(latest) if latest else None
    return flask.jsonify(data_store().get_badge_sum(latest))


@app.route("/api/v2/board/total")
def api2_board_total():
    latest = flask.request.args.get("latest", default=None)
    latest = int(latest) if latest else None
    return flask.jsonify(data_store().get_total_stats(latest))


@app.route("/api/v2/players")
def api2_board_players():
    return flask.jsonify(data_store().get_players())


@app.route("/api/v2/maps")
def api2_maps():
    return flask.jsonify(data_store().get_map_stats())


@app.route("/api/v2/player/<player_id>/kills")
def api2_player_kills(player_id):
    return flask.jsonify(data_store().get_player_kills(player_id))


@app.route("/api/v2/player/<player_id>/deaths")
def api2_player_deaths(player_id):
    return flask.jsonify(data_store().get_player_deaths(player_id))


@app.route("/api/v2/player/<player_id>/badges")
def api2_player_badges(player_id):
    return flask.jsonify(data_store().get_player_badges(player_id))


@app.route("/api/v2/map/size", methods=["POST"])
def api2_map_info():
    if flask.g.user == "admin":
        data_store().set_map_info(
            flask.request.json["map_name"],
            flask.request.json.get("size", None),
            flask.request.json.get("rate", None),
        )
        return "OK"
    return "Bye"


@app.route("/api/v2/upload", methods=["POST"])
def api2_upload():
    if not auth(flask.request.form["token"]):
        return "Bye"

    # TODO this code should be rewritten
    if "file" not in flask.request.files:
        raise Exception("No Files")

    req_file = flask.request.files["file"]
    data = req_file.read().decode("utf-8")
    server_domain = app.config["SERVER_DOMAIN"]

    final_results, errors = manage.process_q3_log(
        server_domain, app.config['RAW_DATA_DIR'], data, 'osp', data_store()
    )

    return flask.jsonify(
        {
            "ACCEPTED_MATCHES": [r.get_summary() for r in final_results],
            "ERRORS": [repr(e) for e in errors],
        }
    )


@app.route("/api/v2/admin/players/merge", methods=["POST"])
def api2_admin_players_merge():
    # TODO PROPER AUTH
    if not auth(flask.request.form["token"]):
        return "Bye"

    source_id = flask.request.form["source_player_id"]
    target_id = flask.request.form["target_player_id"]
    data_store().merge_players(source_id, target_id)
    return "OK"


@app.route("/api/v2/admin/rebuild", methods=["POST"])
def api2_admin_rebuild():
    if not auth(flask.request.form["token"]):
        return "Bye"

    match_count = manage.rebuild_db(
        app.config["RAW_DATA_DIR"], app.config["SERVER_DOMAIN"], data_store
    )

    return "Processed {} matches\n".format(match_count)


@app.route("/api/v2/admin/delete", methods=["POST"])
def api2_admin_delete():
    if not auth(flask.request.form["token"]):
        return "Bye"

    if not flask.request.form["match_guid"]:
        return "Bye"

    data_store().drop_match_info(flask.request.form["match_guid"])
    return "OK"


@app.route("/api/v2/presence/<count>")
def api2_presence(count):
    try:
        count = int(count)
    except ValueError:
        flask.abort(400)

    ds = data_store()
    last_matches = ds.get_matches(count)

    player_ids_per_match = ds.get_match_participants(
        [m["match_guid"] for m in last_matches]
    )

    presence = defaultdict(lambda: 0)
    for match_players in player_ids_per_match.values():
        for player_id in match_players:
            presence[player_id] += 1

    players = ds.get_players(ids=[player_id for player_id in presence.keys()])
    return flask.jsonify({"presence": presence, "players": players})


@app.route("/api/v2/total_stats", methods=['POST'])
def api2_total_stats():
    """
    A little bit hacked endpoint to fetch total stats
    """
    # auth required to prevent DOS as this call is quite expensive
    if not auth(flask.request.form["token"]):
        return "Bye"

    ds = data_store()
    db = ds.db
    matches = list(db.match.find(
        {}, {"_id": 0}
    ))
    result = {}
    result['total_matches'] = len(matches)
    result['total_time_h'] = sum(m['duration'] for m in matches) / (60*60)
    result['total_time_d'] = result['total_time_h'] / 24

    all_players = {
        e['id']: e for e in db.player.find({}, {'_id': 0})
    }

    kills = list(db.kill.find(
        {}, {'_id': 0}
    ))
    non_selfkills = [
        k for k in kills
        if k['killer_id'] != 'q3-world' and k['killer_id'] != k['victim_id']
    ]
    player_stats = list(db.player_stats.find())

    result['total_matches'] = len(matches)
    result['total_time_h'] = sum(m['duration'] for m in matches) / (60*60)
    result['total_time_d'] = result['total_time_h'] / 24
    result['total_kills'] = len(non_selfkills)
    result['total_self_kills'] = len([
        k for k in kills
        if k['killer_id'] == 'q3-world' or k['killer_id'] == k['victim_id']
    ])
    result['world_only_kills'] = len([
        k for k in kills
        if k['killer_id'] == 'q3-world'
    ])
    result['total_players'] = len(all_players)
    result['total_player_mandays'] = (len(player_stats) * 15) / (60 * 8)
    result['total_damage_dealt'] = 0
    result['total_damage_received'] = 0

    player_matches = {}
    for player_stat in player_stats:
        pid = player_stat['player_id']
        if pid not in player_matches:
            player_matches[pid] = {
                'total_games': 0,
                'player_id': pid,
                'total_damage_dealt': 0,
                'total_damage_received': 0,
                'player_name': None
            }
        info = player_matches[pid]
        info['total_games'] += 1
        info['total_damage_dealt'] += player_stat['damage_dealt']
        info['total_damage_received'] += player_stat['damage_taken']
        info['player_name'] = all_players[pid]['name']

        result['total_damage_dealt'] += player_stat['damage_dealt']
        result['total_damage_received'] += player_stat['damage_taken']

    result['top10_played_matches'] = sorted([
        (p['total_games'], p['player_name'])
        for p in player_matches.values()
    ])[-10:]
    result['top10_played_damage_dealt'] = sorted([
        (p['total_damage_dealt'], p['player_name'])
        for p in player_matches.values()
    ])[-10:]
    result['top10_played_damage_received'] = sorted([
        (p['total_damage_received'], p['player_name'])
        for p in player_matches.values()
    ])[-10:]

    player_kd = {}
    for kill in kills:
        kpid = kill['killer_id']
        vpid = kill['victim_id']

        kstat = player_kd.setdefault(
            kpid, {
                'selfkills': 0, 'kills': 0, 'deaths': 0,
                'name': all_players[kpid]['name']
            }
        )
        vstat = player_kd.setdefault(
            vpid, {
                'selfkills': 0, 'kills': 0, 'deaths': 0,
                'name': all_players[vpid]['name']
            }
        )
        kstat['kills'] += 1
        vstat['deaths'] += 1

        if kpid == vpid or kpid == 'q3-world':
            vstat['selfkills'] += 1

    result['top10_kills'] = sorted([
        (k['kills'], k['name']) for k in player_kd.values()
    ])[-10:]
    result['top10_deaths'] = sorted([
        (k['deaths'], k['name']) for k in player_kd.values()
    ])[-10:]

    result['top10_selfkills'] = sorted([
        (k['selfkills'], k['name']) for k in player_kd.values()
    ])[-10:]

    result['unique_players_2+matches'] = len(
        [e for e in player_matches.values() if e['total_games'] > 2]
    )
    result['regular_players_15+matches'] = len(
        [e for e in player_matches.values() if e['total_games'] > 15]
    )
    return flask.jsonify({"matches": result})
